### PR TITLE
Introduce pydantic_v1 compatibility module for pydantic>=2.0.0 support

### PR DESCRIPTION
**Pull request by ringohoffman**
Alternate to #213
Related: https://github.com/microsoft/DeepSpeed/pull/4407

Inspired by LangChain's solution: https://github.com/langchain-ai/langchain/blob/64385c4eae4a4b8c27e73b09018f344f0b791f31/libs/langchain/langchain/pydantic_v1/__init__.py#L14-L17

`mii` can unpin `pydantic`, but instead of using the new `pydantic` v2 features, `mii` can use v2's alias for the v1 API in order to allow greater flexibility for `mii` users by allowing them to use either `pydantic` v1 or v2

### CODE REVIEW

**Comment by ringohoffman** [comment]
@loadams LMK what you think of this solution. I like it because it avoids the pain of upgrading and immediately unblocks people who want to use `deepspeed` but have already upgraded to `pydantic` v2 (me)

**Comment by ringohoffman** [comment]
@microsoft-github-policy-service agree [company="Protopia AI"]

**Comment by ringohoffman** [comment]
@microsoft-github-policy-service agree company="Protopia AI"

**Comment by ringohoffman** [comment]
> Can these workflows be retried? I added the license header, although `scripts/check-license.py` doesn't work on any files for me locally, even ones that are already checked in. I'm on MacOS using bash. Not sure if that that is the problem.

@loadams 🙏 not sure if you were notified of my above comment so just tagging you

**Comment by ringohoffman** [comment]
@loadams it looks like everything is passing. Can you let me know the result of your internal discussion as to whether this PR and https://github.com/microsoft/DeepSpeed/pull/4407 will be acceptable?

Based on what we've discussed, I think the next steps would be to:
1. Merge this PR
2. Rerun workflows for https://github.com/microsoft/DeepSpeed/pull/4407 and merge
3. Revert [`.github/workflows/nv-torch-latest-v100.yaml` Install dependencies URL](https://github.com/microsoft/DeepSpeed-MII/pull/240/files#diff-4411f53253d1981af3cb15521d551a2aa0a071d805e1004c92948e3e7433514aR36)

**Comment by ringohoffman** [comment]
It seems like [`spacy` is also employing a similar approach](https://github.com/explosion/spaCy/commit/245e2ddc2584ff14d0c4a2725e078accd26bd310).

**Comment by loadams** [comment]
> @loadams it looks like everything is passing. Can you let me know the result of your internal discussion as to whether this PR and [microsoft/DeepSpeed#4407](https://github.com/microsoft/DeepSpeed/pull/4407) will be acceptable?
> 
> Based on what we've discussed, I think the next steps would be to:
> 
> 1. Merge this PR
> 2. Rerun workflows for [Introduce pydantic_v1 compatibility module for pydantic>=2.0.0 support DeepSpeed#4407](https://github.com/microsoft/DeepSpeed/pull/4407) and merge
> 3. Revert [`.github/workflows/nv-torch-latest-v100.yaml` Install dependencies URL](https://github.com/microsoft/DeepSpeed-MII/pull/240/files#diff-4411f53253d1981af3cb15521d551a2aa0a071d805e1004c92948e3e7433514aR36)

@ringohoffman - this seems reasonable, we will want feedback from @mrwyattii as well. but your next steps are correct.

**Comment by ringohoffman** [comment]
@mrwyattii do you have a timeline for reviewing this? This is blocking my team from using DeepSpeed due to our dependency on pydantic>=2.0.0.

**Comment by mrwyattii** [comment]
@ringohoffman thank you for the contribution. I'm sorry that this is a blocker for your team. I will prioritize getting this reviewed (along with the accompanying DeepSpeed PR you shared).

File: .github/workflows/nv-torch-latest-v100.yaml [hunk 1]
```diff
@@ -33,7 +33,7 @@ jobs:
 
       - name: Install dependencies
         run: |
-          pip install git+https://github.com/microsoft/DeepSpeed.git
+          pip install git+https://github.com/ringohoffman/DeepSpeed.git@pydantic-compatibility-module
```
Comment (mrwyattii) [inline .github/workflows/nv-torch-latest-v100.yaml:36]:
```suggestion
          pip install git+https://github.com/DeepSpeed.git
```

**Review by mrwyattii — APPROVED** [review]

File: .github/workflows/nv-torch-latest-v100.yaml (hunk 1 above)
Comment (ringohoffman) [reply .github/workflows/nv-torch-latest-v100.yaml]: Shall I apply this now?

File: .github/workflows/nv-torch-latest-v100.yaml [hunk 2]
```diff
@@ -33,7 +33,7 @@ jobs:
 
       - name: Install dependencies
         run: |
-          pip install git+https://github.com/microsoft/DeepSpeed.git
+          pip install git+https://github.com/DeepSpeed.git
```
Comment (mrwyattii) [inline .github/workflows/nv-torch-latest-v100.yaml:36]:
```suggestion
          pip install git+https://github.com/microsoft/DeepSpeed.git
```

**Pull request merged**